text/0182-otlp-remote-parent.md

```diff
@@ -0,0 +1,80 @@
+# Export SpanContext.IsRemote in OTLP
+
+Update OTLP to indicate whether a span's parent is remote.
+
+## Motivation
+
+It is sometimes useful to post-process or visualise only entry-point spans: spans which either have no parent (trace roots), or which have a remote parent.
+For example, the Elastic APM solution highlights entry-point spans (Elastic APM refers to these as "transactions") and surfaces these as top-level operations
+in its user interface.
+
+Currently, the only way entry-point spans can be identified is using (lack of) parent ID, and span kind. Relying on span kind can lead to invalid assumptions,
+particularly with relation to `CONSUMER` messaging spans. Using the [batch receiving example](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving)
+in the messaging semantic conventions, `Span Recv1` should be the only entry point into `Process C`. If we assume `CONSUMER` spans are always entry-point spans,
+then this leads to `Span Proc1` and `Span Proc2` being incorrectly classified as entry-point spans. For messaging spans we might also take into account the
+`messaging.operation` attribute to tell these apart, however `messaging.operation` is not required; and this would not satisfy other scenarios such as actively
+polling a message queue, which would result in a `CONSUMER` span which has a non-remote parent span.
+
+## Explanation
+
+The OTLP encoding for spans has a boolean `parent_span_is_remote` field for identifying whether a span's parent is remote or not.
+All OpenTelemetry SDKs populate this field, and backends may use it to identify a span as being an entry-point span.
+A span can be considered an entry-point span if it has no parent (`parent_span_id` is empty), or if `parent_span_is_remote` is true.
+
+## Internal details
+
+The first part would be to update the trace protobuf, adding a `boolean parent_span_is_remote` field to the
+[`Span` message](https://github.com/open-telemetry/opentelemetry-proto/blob/b43e9b18b76abf3ee040164b55b9c355217151f3/opentelemetry/proto/trace/v1/trace.proto#L84).
+
+OpenTelemetry SDKs already track whether a span's parent is remote in [`SpanContext`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote).
```

Comment:
```suggestion
OpenTelemetry SDKs already track whether a span has been propagated from a remote in [`SpanContext`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote).
```

The SDKs track whether a "span is remote", not specifically the span's parent. When creating a new span, then the tracer can check whether the parent span is remote and fill in the proposed field. The spirit of this line, that it is just a one line change in SDKs to fill the data, is correct but want to make sure the semantics of `is_remote` aren't confused

Reply:
Maybe it's just my pedantic brain but I find that a bit confusing: spans aren't propagated, span context is. Let me see if we can come to a compromise...

The [docs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote) say:

> An API called IsRemote, that returns a boolean value, which is true if the SpanContext was propagated from a remote parent

How about I just repeat that here?

```suggestion
[`SpanContext.IsRemote`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isremote) identifies whether span context has been propagated from a remote parent.
```

Reply:
OTel went with the approach of `NonRecordingSpan` to hold a propagated span context

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#wrapping-a-spancontext-in-a-span

It's true that only a context is propagated but the SDKs happen to treat things a little differently

Reply:
That being said your correction seems fine too